Add shape level inflate method

DusKing1 opened the pull request:
Fix #168 . Please kindly review this and just ask me to change code that you don't want it to be that way. Really appreciate if you can find your time to do the review job!

fontanf commented:
Thank you. That looks good overall.

* The inflate function should take shape holes as parameters as well (`const std::vector<Shape>& holes`) and deflate them.

* Then you should remove the `GenralizedTrapezoid::inflate` method completely and replace the other calls by shape inflate and then trapezoidation.

* Could you add some comments in the inflate function so that I can understand and maintain it?

* Could you add a couple of unit tests in this file for example https://github.com/fontanf/packingsolver/blob/master/test/irregular/shape_test.cpp

DusKing1 commented:
Sure, lemme do that.

DusKing1 commented:
Urghhh, making this such a huge change was not my original intention.... Please let me know if you still need me to change the code.

fontanf commented:
I would prefer if you don't include the refactoring, it makes the PR harder to read.

The more so, the changes broke multiple tests, including several not related to minimum distances:

```
 	  2 - IrregularTrapezoid.Intersection1 (Failed)
	  3 - IrregularTrapezoid.Intersection2 (Failed)
	  4 - IrregularTrapezoid.Intersection3 (Failed)
	 12 - IrregularShape.InflateWithHoles (Failed)
	 41 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/rectangles_non_guillotine.json" (Failed)
	 42 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/direction_y.json" (Failed)
	 43 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-06-25.json" (Failed)
	 44 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/non_rectangular_bin.json" (Failed)
	 45 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/polygon_with_hole.json" (Failed)
	 46 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-07-01.json" (Failed)
	 47 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/rotations.json" (Failed)
	 48 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/multiple_bins.json" (Failed)
	 49 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/trapezoidal_insertions.json" (Failed)
	 50 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/trapezoidal_insertions_2.json" (Failed)
	 51 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/trapezoidal_insertions_3.json" (Failed)
	 52 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/trapezoidal_insertions_4.json" (Failed)
	 53 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-08-05.json" (Failed)
	 54 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-08-08_sub_1.json" (Failed)
	 55 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-08-08_sub_2.json" (Failed)
	 56 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-08-08_sub_3.json" (Failed)
	 57 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-08-08.json" (Failed)
	 58 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-08-13.json" (Failed)
	 59 - Irregular/AlgorithmTest.Algorithm/"data/irregular/tests/mirror.json" (Failed)
	 60 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-08-25_1.json" (Failed)
	 61 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-08-25_2.json" (Failed)
	 62 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-08-25_3.json" (Failed)
	 63 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-08-25_4.json" (Failed)
	 64 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-09-11.json" (Failed)
	 65 - Irregular/AlgorithmTest.Algorithm/"data/irregular/users/2024-09-21.json" (Failed)
```

If you remove the refactoring and the failing tests are still there, I could look at it as well.

Could you also remove the `holes` attribute you added in the Shape structure?

fontanf commented:
Additionally, inflating/deflating a polygon may lead to self-intersecting inflated/deflated polygons. There are functions in the library to post-process self-intersecting polygons here https://github.com/fontanf/packingsolver/blob/master/src/irregular/shape_self_intersections_removal.hpp

You can see how they are used in the polygon simplification algorithm https://github.com/fontanf/packingsolver/blob/master/src/irregular/shape_simplification.cpp

And that would be relevant to have unit tests for these cases.

DusKing1 commented:
So lemme reset the head to 87e77e429092337ba0739d841acb6b6cc0cf26b9?

fontanf commented:
Or you can do a commit to undo the changes. It will be squashed in the end anyway.

DusKing1 commented:
Reset back to 87e77e42, what do you want me to change?

fontanf commented:
Thank you. You can do what I mentioned in the first message https://github.com/fontanf/packingsolver/pull/169#issuecomment-2737252249 and in this other message https://github.com/fontanf/packingsolver/pull/169#issuecomment-2738141598

Please run the tests locally
```
cd build/test/
ctest --parallel --output-on-failure
```

And check that the current tests and your new tests are working

Please let me know if you get stuck to fix some tests. I could take a look then.

fontanf commented:
Yes, that's it

DusKing1 commented:
I still have FAILED test:
```
The following tests FAILED:
          1 - PackingSolver_boxstacks_test_NOT_BUILT (Not Run)
         39 - IrregularShapeInflate.BasicInflate (Failed)
         40 - IrregularShapeInflate.InflateWithHoles (Failed)
         41 - IrregularShapeInflate.SelfIntersectingInflate (Failed)
         42 - IrregularShapeInflate.Deflate (Failed)
         43 - IrregularShapeInflate.TinyShape (Failed)
         44 - IrregularShapeInflate.LargeInflation (Failed)
         46 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\rectangles_non_guillotine.json" (Failed)
         47 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\direction_y.json" (Failed)
         48 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-06-25.json" (Failed)
         49 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\non_rectangular_bin.json" (Failed)
         50 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\polygon_with_hole.json" (Failed)
         51 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-07-01.json" (Failed)
         52 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\rotations.json" (Failed)
         53 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\multiple_bins.json" (Failed)
         54 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\trapezoidal_insertions.json" (Failed)
         55 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\trapezoidal_insertions_2.json" (Failed)
         56 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\trapezoidal_insertions_3.json" (Failed)
         57 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\trapezoidal_insertions_4.json" (Failed)
         58 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-05.json" (Failed)
         59 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-08_sub_1.json" (Failed)
         60 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-08_sub_2.json" (Failed)
         61 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-08_sub_3.json" (Failed)
         62 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-08.json" (Failed)
         63 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-13.json" (Failed)
         64 - Irregular/AlgorithmTest.Algorithm/"data\irregular\tests\mirror.json" (Failed)
         65 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-25_1.json" (Failed)
         66 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-25_2.json" (Failed)
         67 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-25_3.json" (Failed)
         68 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-25_4.json" (Failed)
         69 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-09-11.json" (Failed)
         70 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-09-21.json" (Failed)
Errors while running CTest
```
Can you take a look?

fontanf commented:
Your new tests only call the code you added:
```
         39 - IrregularShapeInflate.BasicInflate (Failed)
         40 - IrregularShapeInflate.InflateWithHoles (Failed)
         41 - IrregularShapeInflate.SelfIntersectingInflate (Failed)
         42 - IrregularShapeInflate.Deflate (Failed)
         43 - IrregularShapeInflate.TinyShape (Failed)
         44 - IrregularShapeInflate.LargeInflation (Failed)
```

Please fix these first.

DusKing1 commented:
It seems that the compute_intersections can't handle CircularArc properly. Do you want me to modify it so to support the CircularArc element?

fontanf commented:
Even if the input allows defining circular arcs, the solver doesn't handle them (yet) and there would be multiple other functions to update to do so.

But I actually think it's good to have the inflate function generating the rounded corners.

For now, we'll break them into multiple line segments after the inflate.

So yes, if you feel you can update this function to handle circular arcs, I think it's the best thing to do. I can handle this as well, but probably in at least a few days.

fontanf commented:
Implementing full support for circular arc is a big and not that obvious task.

If you can have an inflate function using circular arcs which is working and well tested, that would be very good for this PR.

Then I could take care of finishing its integration, or you could do it as well if you're more in a hurry

DusKing1 commented:
Bed time hehe. I'll just commit them here. It will be great if you can take a look into the new inflate function which is still WIP.

Another question, what about just implementing the Minkowski sum as the inflate function instead of struggling with such an approximation which I'm writing?

fontanf commented:
Choose an approach that you can make work reliably.

I refactored and cleaned the shape code. It should be easier to write tests the same way as this one for example now:

https://github.com/fontanf/packingsolver/blob/855d304e8af2ffba4449f6c0cbd9a2c508345da2/test/irregular/shape_test.cpp#L8-L17

I also updated the `build_shape` (previously `build_polygon_shape`) function to be able to use it to generate shapes with circular arcs:

https://github.com/fontanf/packingsolver/blob/855d304e8af2ffba4449f6c0cbd9a2c508345da2/include/packingsolver/irregular/shape.hpp#L229-L244

fontanf commented:
@bbeaulant thank you for your vigilance. I pushed a fix for this on the current dev branch.

fontanf commented:
@DusKing1 could you rebase your branch and move the inflate function and its subfunctions to new `shape_inflate.{hpp,cpp}` files?

DusKing1 commented:
@fontanf yes for sure, lemme commit my latest work and rebase then do the move. I have less fail tests now.

DusKing1 commented:
I think now the inflate function is ready for your review. Here are the remaining failed tests that I don't know how to fix:

```
The following tests FAILED:
         53 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-07-01.json" (Failed)
         61 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-08_sub_1.json" (Failed)
         63 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-08_sub_3.json" (Failed)
         64 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-08.json" (Failed)
         65 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-13.json" (Failed)
         67 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-25_1.json" (Failed)
         68 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-25_2.json" (Failed)
         69 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-25_3.json" (Failed)
         70 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-08-25_4.json" (Failed)
         71 - Irregular/AlgorithmTest.Algorithm/"data\irregular\users\2024-09-11.json" (Failed)
```

Please lemme know if you have further change requests.

fontanf commented:
Thank you for the update.

In the tests, could you check the whole shape instead of the just the min and max? I added a `near` function to compare 2 shapes. I also changed some tests to parametrized tests. That would be relevant here as well. For example

https://github.com/fontanf/packingsolver/blob/6ea89b0edc9aec40aadb24f6b4db516124ca0130/test/irregular/shape_test.cpp#L40-L92

Could you translate Chinese comments into English?

To compare to 0, could you use `equal(value, 0.0)` instead of comparing to 1e-6?

```
                    // For small gaps, prefer line segment connectors
                    if (gap_distance < 0.1) {
```

I'm not a big fan of this, we cannot assume that 0.1 is small. All lengths could be around `0.01`.

There are functions to compute angles in the library that you can use directly:

https://github.com/fontanf/packingsolver/blob/6ea89b0edc9aec40aadb24f6b4db516124ca0130/src/irregular/shape.cpp#L120-L137

```
    // Remove duplicate points or degenerate elements
    if (inflated_shape.elements.size() > 1) {
        std::vector<ShapeElement> cleaned_elements;
        cleaned_elements.reserve(inflated_shape.elements.size());

        for (const ShapeElement& element : inflated_shape.elements) {
            if (euclidean_distance(element.start, element.end) > 1e-6) {
                cleaned_elements.push_back(element);
            }
        }

        inflated_shape.elements = cleaned_elements;
    }
```

Could the `clean_shape` function already available in the library replace this piece of code?

That great to have an `is_point_inside_shape` function. Is it strictly inside or not? could you explicit it in the name? And inside the function, could you use the `equal`, `strictly_lesser` and `strictly_greater` functions to do the comparisons?

Could you undo the changes in `branching_scheme.cpp` and `trapezoid.hpp`? Because of the circular arcs in the inflated shapes, the integration will need to be a bit more sophisticated. But that should provide better results in the end.

DusKing1 commented:
Bed time hehe, I need to check the near() and learn how to use it tomorrow.

For now I update the code to meet all your request other than using the near() in tests. It will be great if you can review them. Thanks!

fontanf commented:
Thank you. I look forward to the test update to look at them and get convinced of the inflate function.

Here are the remaining comments:

Can you rebase your branch to the latest master?

I moved the `shape.hpp` file into `include/packingsolver`, the file src/shape.hpp still exists in your branch. Can you move your changes into the `include/packingsolver/shape.hpp` file?

Could you remove the unused includes from `shape.hpp`
```
#include "irregular/shape_closure.hpp"
#include "irregular/shape_self_intersections_removal.hpp"
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits>
#include <vector>
```
And include the files only in the files where the included functions are actually used?

Could you remove from `src/irregular/shape_closure.hpp` the function declarations of functions which are not called outside of `src/irregular/shape_closure.cpp`?

Do you still need the changes in `src/irregular/shape_self_intersections_removal.cpp`? If no, could you undo them?

DusKing1 commented:
Hi @fontanf , I've updated the code per your request. Now this PR only contains add lines. Please find time to review it if possible.

I'll continue working on adding proper test code for this inflate function, just need to squeeze out more spare time which I lack of hehe :)

BTW, it would be great if you can just use the GitHub's code review feature to comment directly on my code, by `Files changed` then select the code or codeblock you wanna comment on. It will also help me to catch up easier.

DusKing1 commented:
Hi @fontanf , I've met some precision issue. do you know how to fix this?

```
32: Actual shape elements:
32: Element 0: LineSegment start (-2, 10) end (-2, 0)
32: Element 1: CircularArc start (-2, 0) end (0, -2) center (0.414214, 0.414214) anticlockwise
32: Element 2: LineSegment start (0, -2) end (10, -2)
32: Element 3: CircularArc start (10, -2) end (12, 0) center (9.58579, 0.414214) anticlockwise
32: Element 4: LineSegment start (12, 0) end (12, 10)
32: Element 5: CircularArc start (12, 10) end (10, 12) center (9.58579, 9.58579) anticlockwise
32: Element 6: LineSegment start (10, 12) end (0, 12)
32: Element 7: CircularArc start (0, 12) end (-2, 10) center (0.414214, 9.58579) anticlockwise
32:
32: Expected shape elements:
32: Element 0: LineSegment start (-2, 10) end (-2, 0)
32: Element 1: CircularArc start (-2, 0) end (0, -2) center (0, 0) anticlockwise
32: Element 2: LineSegment start (0, -2) end (10, -2)
32: Element 3: CircularArc start (10, -2) end (12, 0) center (10, 0) anticlockwise
32: Element 4: LineSegment start (12, 0) end (12, 10)
32: Element 5: CircularArc start (12, 10) end (10, 12) center (10, 10) anticlockwise
32: Element 6: LineSegment start (10, 12) end (0, 12)
32: Element 7: CircularArc start (0, 12) end (-2, 10) center (0, 10) anticlockwise
```

fontanf commented:
The difference seems way to large to be a numerical issue.

fontanf commented:
This formula is probably wrong

https://github.com/DusKing1/packingsolver/blob/f5ef6a697a2e4367f39f6e7d72999925b3b72b25/src/irregular/shape_closure.cpp#L353-L356

I don't have time to investigate more right now.

DusKing1 commented:
Thanks fore the hint, I'll recheck my formula.

fontanf commented on src/irregular/shape_inflate.cpp:
```diff
@@ -0,0 +1,340 @@
+            // This is a simplified check, an actual implementation would need more precise geometric calculations
+            if (std::abs(std::abs(dot_product) - 1.0) < 0.1) {
+                // Calculate the arc radius
+                LengthDbl radius = std::sqrt(
+                    std::pow(arc_element.start.x - arc_element.center.x, 2) + 
+                    std::pow(arc_element.start.y - arc_element.center.y, 2)
```
Use the `distance` function from the library

https://github.com/fontanf/packingsolver/blob/aa35eb9bff4883dfdfe8e4a32aead2942d24bd2e/include/packingsolver/irregular/shape.hpp#L66-L68

fontanf replied:
And update the other places in the code where the distance function should be used

DusKing1 replied:
Done.

fontanf commented on src/irregular/shape_inflate.cpp:
```diff
@@ -0,0 +1,340 @@
+        const ShapeElement& element = original_shape.elements[i];
+        ShapeElement inflated_element = offset_element(element, value);
+        
+        std::cout << "Original element " << i << ": ";
+        if (element.type == ShapeElementType::LineSegment) {
+            std::cout << "Line Segment";
+        } else {
+            std::cout << "Circular Arc";
+        }
+        std::cout << " start(" << element.start.x << ", " << element.start.y 
+                  << "), end(" << element.end.x << ", " << element.end.y << ")";
+        
+        if (element.type == ShapeElementType::CircularArc) {
+            std::cout << ", center(" << element.center.x << ", " << element.center.y 
+                      << "), " << (element.anticlockwise ? "counterclockwise" : "clockwise");
+        }
+        std::cout << std::endl;
```
There is a `ShapeElement:to_string` method

https://github.com/fontanf/packingsolver/blob/aa35eb9bff4883dfdfe8e4a32aead2942d24bd2e/src/irregular/shape.cpp#L155-L169

fontanf commented on src/irregular/shape_inflate.cpp:
```diff
@@ -0,0 +1,340 @@
+    std::cout << "\n==== Debug: Creating Shape Inflation Mapping ====" << std::endl;
+    std::cout << "Original shape element count: " << original_shape.elements.size() << std::endl;
+    
+    for (size_t i = 0; i < original_shape.elements.size(); ++i) {
+        const ShapeElement& element = original_shape.elements[i];
```
Replace by

```cpp
    for (ElementPos element_pos = 0; element_pos < (ElementPos)original_shape.elements.size(); ++element_pos) {
        const ShapeElement& element = original_shape.elements[element_pos];
```

DusKing1 replied:
So it should be:
```suggestion
    for (ElementPos element_pos = 0; element_pos < (ElementPos)original_shape.elements.size(); ++element_pos) {
        const ShapeElement& element = original_shape.elements[element_pos];
```

fontanf commented on src/irregular/shape_inflate.cpp:
```diff
@@ -0,0 +1,340 @@
+            // Calculate dot product to determine the angle
+            LengthDbl dot_product = dir1_x * dir2_x + dir1_y * dir2_y;
+            
+            // If the two segments are almost parallel (dot product close to 1 or -1), and the inflation value is large enough
+            // This is a simplified check, an actual implementation would need more precise geometric calculations
+            if (std::abs(std::abs(dot_product) - 1.0) < 0.1) {
```
This doesn't give a lot of confidence

DusKing1 replied:
This no more exists

fontanf commented on src/irregular/shape_inflate.cpp:
```diff
@@ -0,0 +1,340 @@
+        LengthDbl length = std::sqrt(dx * dx + dy * dy);
+        
+        // Avoid division by zero
+        if (length > 0) {
```
What happens in the other case?

DusKing1 replied:
If the length of the segment is ==0, then this won't process it, in later func call in `is_degenerate_element()` it (length==0) will be degenerated.

fontanf commented on src/irregular/shape_inflate.cpp:
```diff
@@ -0,0 +1,340 @@
+            new_radius = radius - value;
+        }
+        
+        // Skip if radius becomes too small
+        if (new_radius <= 0.01) {
+            return new_element; // Return empty/invalid element
+        }
```
0.01 is not too small. This shouldn't be checked

DusKing1 replied:
```suggestion
```

fontanf commented on src/irregular/shape_inflate.cpp:
```diff
@@ -0,0 +1,340 @@
+        }
+        
+        // Calculate the original arc's start and end angles
+        LengthDbl start_angle = angle_radian({element.start.x - element.center.x, element.start.y - element.center.y});
+        LengthDbl end_angle = angle_radian({element.end.x - element.center.x, element.end.y - element.center.y});
```
The brackets {} are not needed

https://github.com/fontanf/packingsolver/blob/aa35eb9bff4883dfdfe8e4a32aead2942d24bd2e/src/irregular/shape.cpp#L129-L137

DusKing1 replied:
If remove the bracket then it report compiler error

fontanf replied:
Does this work?
```cpp
angle_radian(element.start - element.center)
```

fontanf replied:
Also, the type for angles should be `Angle` instead of `LengthDbl`

DusKing1 replied:
So it should be:
```suggestion
        LengthDbl start_angle = angle_radian({element.start.x - element.center.x, element.start.y - element.center.y});
        LengthDbl end_angle = angle_radian({element.end.x - element.center.x, element.end.y - element.center.y});
        Angle start_angle = angle_radian(element.start - element.center);
        Angle end_angle = angle_radian(element.end - element.center);
```

fontanf commented on src/irregular/shape_inflate.cpp:
```diff
@@ -0,0 +1,340 @@
+    return new_element;
+}
+
+// Helper function to check if an element is degenerate (too small)
+bool is_degenerate_element(const ShapeElement& element)
+{
+    if (element.type == ShapeElementType::LineSegment) {
+        LengthDbl dx = element.end.x - element.start.x;
+        LengthDbl dy = element.end.y - element.start.y;
+        LengthDbl length = std::sqrt(dx * dx + dy * dy);
+        return equal(length, 0.0);
+    } else if (element.type == ShapeElementType::CircularArc) {
+        LengthDbl radius = std::sqrt(
+            std::pow(element.start.x - element.center.x, 2) + 
+            std::pow(element.start.y - element.center.y, 2)
+        );
+        
+        // If the radius is very small, consider it degenerate
+        if (equal(radius, 0.0)) {
+            return true;
+        }
+        
+        // If start and end points are too close, consider it degenerate
+        LengthDbl dx = element.end.x - element.start.x;
+        LengthDbl dy = element.end.y - element.start.y;
+        LengthDbl chord_length = std::sqrt(dx * dx + dy * dy);
+        return equal(chord_length, 0.0);
+    }
+    
+    return false;
+}
```
The shapes passed to the inflate function will have passed through the `clean_shape` function before. We may consider that it doesn't contain such degenerate element, so this function is not needed

DusKing1 replied:
(Please fogive me about my ungly hand-drawing)
![image](https://github.com/user-attachments/assets/dfb53738-a2b6-4fb7-b793-b5ca03ff8702)
Considering this case, then blue elements won't be on the inflated outline after the inflation, so it will be degenerated...
Or, do you mean that the `clean_shape` function can handle this case and delete the blue elements before calling the `inflate()`? I haven't check it closely..

DusKing1 replied:
Sure, I'll add a new test for this case.

DusKing1 replied:
Sorry, I misremembered the meaning of this function. It is actually specific to the deflation of arcs. When the deflation value is greater than the arc radius, the arc should degenerate.

fontanf commented on src/irregular/shape_inflate.cpp:
```diff
@@ -0,0 +1,340 @@
+namespace irregular
+{
+
+// Check if an arc is covered by adjacent elements
+bool is_arc_covered_by_adjacent_elements(
+        const ShapeElement& prev_element,
+        const ShapeElement& arc_element,
+        const ShapeElement& next_element,
+        LengthDbl value)
```
I don't see this function called anywhere. Remove it if it is not used.

DusKing1 replied:
Removed

fontanf commented on src/irregular/shape_closure.cpp:
```diff
@@ -0,0 +1,504 @@
+namespace irregular
+{
+
+LengthDbl euclidean_distance(const Point& p1, const Point& p2) {
+    LengthDbl dx = p2.x - p1.x;
+    LengthDbl dy = p2.y - p1.y;
+    return std::sqrt(dx * dx + dy * dy);
+}
```
Duplicate with the `distance` function

https://github.com/fontanf/packingsolver/blob/aa35eb9bff4883dfdfe8e4a32aead2942d24bd2e/src/irregular/shape.cpp#L37-L42

fontanf commented on src/irregular/shape_closure.cpp:
```diff
@@ -0,0 +1,504 @@
+        //          << ", start=(" << next.start.x << "," << next.start.y 
+        //          << "), end=(" << next.end.x << "," << next.end.y << ")" << std::endl;
+        
+        // Check if a connector element is needed
+        LengthDbl gap_distance = euclidean_distance(current.end, next.start);
+        //std::cout << "  Gap distance: " << gap_distance << std::endl;
+        
+        if (!equal(gap_distance, 0.0)) {
```
That's probably not a good idea to check if two elements are equal this way. To be "equal" to 0.0, a number needs to be very small. If the coordinates are large, we might want to consider them equal but their difference wouldn't be "equal" to 0.0. You should use `equal(current.end.x, next.start.x) && equal(current.end.y, next.start.y)` or directly the `near` function (which I'll probably rename `equal` later)

https://github.com/fontanf/packingsolver/blob/aa35eb9bff4883dfdfe8e4a32aead2942d24bd2e/include/packingsolver/irregular/shape.hpp#L310-L315

DusKing1 replied:
Do you want me to use `near()` or, do you want me to rebase my branch after you renamed that to equal?

fontanf replied:
I pushed the rename. You can rebase and use the `equal` function.

DusKing1 replied:
Done

DusKing1 commented:
Thanks for your careful reviewing, I fixed some of them. For those (unfixed and uncommented), I'll back to check them and reply you thoughtfully. Now the code can just pass all tests.

fontanf commented on src/irregular/shape_closure.hpp:
I think the content of `shape_closure.hpp/cpp` is short enough to include it back into `shape_inflate.hpp/cpp` so we have everything related to shape inflation at the same place

DusKing1 replied:
Make sense to me, I'll merge it into that file.

fontanf commented on src/irregular/shape_inflate.cpp:
```diff
@@ -0,0 +1,274 @@
+// Helper function to check if an element is degenerate (too small)
+bool is_degenerate_element(const ShapeElement& element)
+{
+    if (element.type == ShapeElementType::LineSegment) {
+        LengthDbl length = distance(element.start, element.end);
+        return equal(length, 0.0);
+    } else if (element.type == ShapeElementType::CircularArc) {
+        LengthDbl radius = distance(element.center, element.start);
+        
+        // If the radius is very small, consider it degenerate
+        if (equal(radius, 0.0)) {
+            return true;
+        }
+        
+        // If start and end points are too close, consider it degenerate
+        LengthDbl chord_length = distance(element.start, element.end);
+        return equal(chord_length, 0.0);
```
Here as well, it would be better to check equality between end points than checking if their distance is equal to 0:

```cpp
equal(element.start, element.end)
...
equal(element.center, element.start)
...
equal(element.start, element.end)
```

DusKing1 replied:
Done

DusKing1 commented:
Updated inflate logic when handling corner points:
- When inflating
    - convex points need to add an arc connector
    - concave points add nothing
- When deflating
    - convex points add nothing
    - concave points add an arc connector

Also, updated the case when multiple continuous degenerate elements detected, the arc logic is verified by paper proof :)

Next is to enhance the func of removing self-intersecting sub-elements to handle arcs properly

DusKing1 commented:
I feel like I'm going down a rabbit hole:
If I want to handle the self-intersection of my current `inflate()` function, it seems to involve very complicated multi-tree data structure design and depth-first search algorithm design.

A simpler method seems to be to directly cut the original graph into trapezoids(simply consider arcs as linesegments when cut) without any concave points, then to inflate them and then re-merge the graph (graphical Boolean operation).

Can you give me some guidance? @fontanf

fontanf commented:
Could you illustrate why the current `remove_self_intersections` function doesn't work in this case?

Did you try to see how this is handled in clipper?

DusKing1 commented:
> Could you illustrate why the current `remove_self_intersections` function doesn't work in this case?

The current `remove_self_intersections` will call `compute_graph`, which can't handle the `CircularArc` case.

> Did you try to see how this is handled in clipper?

I haven't checked Clipper2 very deeply, but I think it simply assumes the input shape is a polygon and no arc. It approximates arc by using series of relatively short straight line segments.

fontanf commented:
> The current `remove_self_intersections` will call `compute_graph`, which can't handle the `CircularArc` case. I haven't checked it deeply cuz I don't understand the math behind it very well.

At first glance, I would say that filling the `compute_intersections` function to handle circular arcs should be enough to make the `remove_self_intersections` work with circular arcs. If you don't see how to implement it, I can try to look at it next week

DusKing1 commented:
> At first glance, I would say that filling the compute_intersections function to handle circular arcs should be enough to make the remove_self_intersections work with circular arcs.

So it should be something like what I did in [4525846](https://github.com/fontanf/packingsolver/pull/169/commits/4525846e606a4c50e0a328281fbea1384e3db718)?

fontanf commented:
It looks like. I didn't check the details of your implementation. Did you find cases where then the `remove_self_intersections` doesn't work?

If you need to test the `compute_intersection` function independently, you can put its declaration into the header file to make it possible to write unit tests.

DusKing1 commented:
OK. lemme test that firstly :)

fontanf commented:
I'm interested in a `compute_intersections` function that can handle circular arcs on its own.

If your implementation is working, could you do a new pull request for this feature only?

Actually, I would like to distinguish between `compute_intersections_and_contact_points` (the current behavior) and `compute_strict_intersections`. But I could do this myself after.

DusKing1 commented:
> I'm interested in a `compute_intersections` function that can handle circular arcs on its own.
> 
> If your implementation is working, could you do a new pull request for this feature only?

I also think the changes to the `compute_intersextions` are beyond the scpoe of this PR, so I'll create a seperate PR for it.

But I'm struggling make proper test code for it.. It would be nice if you can input some help on it.

> Actually, I would like to distinguish between `compute_intersections_and_contact_points` (the current behavior) and `compute_strict_intersections`. But I could do this myself after.

Sure thing, I'll then do the rebase after that.

fontanf commented:
> But I'm struggling make proper test code for it.. It would be nice if you can input some help on it.

What are you struggling with? Here is another test using circular arcs and line segments that you can start from:
https://github.com/fontanf/packingsolver/blob/9bba8e66d71a47a8c2c64d19fda03fa9116050a0/test/irregular/shape_test.cpp#L64-L131

fontanf commented:
Hi,

I decided to move the shape package in a separate repository since I need it for other applications.

https://github.com/fontanf/shape

The shape inflate function should now be added there. I close this pull request. Could you move your code to the new repository?

I apologize for the inconvenience. The development of the library is quite active right now and some things change quickly

DusKing1 commented:
Ok, I'll check that repo later, and maybe I'll need some guidance from you on how to setup that repo to continue development.
Just had some progress on the math these days, I think I'll made this inflate method more robust and reasonable.

fontanf commented:
> Ok, I'll check that repo later, and maybe I'll need some guidance from you on how to setup that repo to continue development.

The repository is very similar to the previous shape code of the packing library. But do not hesitate if you have question

> Just had some progress on the math these days, I think I'll made this inflate method more robust and reasonable.

Great news
